File: backend/app/schemas/runtime_settings.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


DraftRewriteIntensity = Literal["light", "moderate", "strong"]
DraftRewriteTone = Literal["polite", "professional", "friendly"]
DraftRewriteFormality = Literal["natural", "balanced", "formal"]
DraftRewriteLength = Literal["shorter", "default", "more_detailed"]
DraftRewriteSpecificity = Literal["concise", "balanced", "detailed"]
DraftTemplatePreservation = Literal["structure_first", "balanced", "content_first"]


class RuntimeSettingsRead(BaseModel):
    match_analysis_job_worker_count: int
    match_analysis_job_item_concurrency: int
    match_analysis_job_interval_seconds: int
    crawler_worker_count: int
    crawler_profile_enrichment_concurrency: int
    crawler_host_concurrency: int
    draft_max_tokens: int
    batch_draft_generation_concurrency: int
    draft_rewrite_intensity: DraftRewriteIntensity
    draft_rewrite_tone: DraftRewriteTone
    draft_rewrite_formality: DraftRewriteFormality
    draft_rewrite_length: DraftRewriteLength
    draft_rewrite_specificity: DraftRewriteSpecificity
    draft_template_preservation: DraftTemplatePreservation
    updated_at: datetime


class RuntimeSettingsUpdate(BaseModel):
    match_analysis_job_worker_count: int = Field(ge=1, le=8)
    match_analysis_job_item_concurrency: int = Field(ge=1, le=20)
    match_analysis_job_interval_seconds: int = Field(ge=1, le=300)
    crawler_worker_count: int = Field(ge=1, le=8)
    crawler_profile_enrichment_concurrency: int = Field(ge=1, le=20)
    crawler_host_concurrency: int = Field(ge=1, le=8)
    draft_max_tokens: int = Field(ge=256, le=32000)
    batch_draft_generation_concurrency: int = Field(ge=1, le=20)
    draft_rewrite_intensity: DraftRewriteIntensity
    draft_rewrite_tone: DraftRewriteTone
    draft_rewrite_formality: DraftRewriteFormality
    draft_rewrite_length: DraftRewriteLength
    draft_rewrite_specificity: DraftRewriteSpecificity
    draft_template_preservation: DraftTemplatePreservation

File: frontend/src/lib/api/runtimeSettings.ts
import { apiFetch } from "@/lib/api/client";

export type DraftRewriteIntensity = "light" | "moderate" | "strong";
export type DraftRewriteTone = "polite" | "professional" | "friendly";
export type DraftRewriteFormality = "natural" | "balanced" | "formal";
export type DraftRewriteLength = "shorter" | "default" | "more_detailed";
export type DraftRewriteSpecificity = "concise" | "balanced" | "detailed";
export type DraftTemplatePreservation = "structure_first" | "balanced" | "content_first";

export interface RuntimeSettingsDTO {
  match_analysis_job_worker_count: number;
  match_analysis_job_item_concurrency: number;
  match_analysis_job_interval_seconds: number;
  crawler_worker_count: number;
  crawler_profile_enrichment_concurrency: number;
  crawler_host_concurrency: number;
  draft_max_tokens: number;
  draft_rewrite_intensity: DraftRewriteIntensity;
  draft_rewrite_tone: DraftRewriteTone;
  draft_rewrite_formality: DraftRewriteFormality;
  draft_rewrite_length: DraftRewriteLength;
  draft_rewrite_specificity: DraftRewriteSpecificity;
  draft_template_preservation: DraftTemplatePreservation;
  updated_at: string;
}

export type RuntimeSettingsUpdateDTO = Omit<RuntimeSettingsDTO, "updated_at">;

export const defaultDraftRewritePreferences = {
  draft_rewrite_intensity: "moderate",
  draft_rewrite_tone: "polite",
  draft_rewrite_formality: "balanced",
  draft_rewrite_length: "default",
  draft_rewrite_specificity: "balanced",
  draft_template_preservation: "structure_first",
} satisfies Pick<
  RuntimeSettingsDTO,
  | "draft_rewrite_intensity"
  | "draft_rewrite_tone"
  | "draft_rewrite_formality"
  | "draft_rewrite_length"
  | "draft_rewrite_specificity"
  | "draft_template_preservation"
>;

export const getRuntimeSettings = () =>
  apiFetch<RuntimeSettingsDTO>("/api/runtime-settings");

export const updateRuntimeSettings = (payload: RuntimeSettingsUpdateDTO) =>
  apiFetch<RuntimeSettingsDTO>("/api/runtime-settings", {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
